nominal/client.py

```diff
@@ -0,0 +1,481 @@
+        ).collect()
+        return (bounds["min"].item(), bounds["max"].item())
+
+    # todo - merge this with the existing _upload_s3 function, it's mostly copy-pasted wholesale
+    def _upload_file_s3(self):
+        if hasattr(self, "s3_path"):
+            print(f"already uploaded dataset {self.csv_path} to {self.s3_path}")
+            return
+
+        self.filename = os.path.basename(self.csv_path)
+
+        csv_file = open(self.csv_path, "rb")
+        csv_buffer_size_bytes = os.stat(self.csv_path).st_size
+
+        print(
+            "\nUploading: [bold green]{0}[/bold green]\nto {1}\n = {2} bytes".format(
+                self.filename, get_base_url, csv_buffer_size_bytes
+            )
+        )
+
+        # Make POST request to upload data file to S3
+        resp = requests.post(
+            url=f"{get_base_url}/upload/v1/upload-file?fileName={self.filename}",
+            data=csv_file.read(),
+            params={"sizeBytes": csv_buffer_size_bytes},
+            headers={"Authorization": f"Bearer {TOKEN}", "Content-Type": "application/octet-stream"},
+        )
+
+        if resp.status_code == 200:
+            self.s3_path = resp.text.strip('"')
+            print("\nUpload to S3 successful. S3 bucket:", self.s3_path)
+        else:
+            print("\n{0} error during upload to S3:\n".format(resp.status_code), resp.json())
+
+    def _init_sess(self):
+        if hasattr(self, "sess"):
+            return
+        self.sess = chs.Session()
+        self.sess.query("CREATE DATABASE IF NOT EXISTS nominal")
+        self.sess.query("""
+        CREATE TABLE IF NOT EXISTS nominal.dataset_float64 (series String,
+        timestamp DateTime64(9) CODEC(Delta, ZSTD), value Float64)
+        ENGINE=ReplacingMergeTree() ORDER BY (series, timestamp)
+        """)
+        self.sess.query("""
+        CREATE TABLE IF NOT EXISTS nominal.dataset_string (series String,
+        timestamp DateTime64(9) CODEC(Delta, ZSTD), value String)
+        ENGINE=ReplacingMergeTree() ORDER BY (series, timestamp)
+        """)
+
+    def run(self, name: str):
+        (ts_min, ts_max) = self._bounds()
+        return Run(name, [self], ts_min, ts_max)
+
+    def _series(self, series: str):
+        if series not in self.col_series:
+            return None
+        return self.col_series[series]
+
+
+class Run:
+    def __init__(self, name: str, sources: list[Dataset], start: int, end: int):
+        self.name = name
+        self.sources = sources
+        self.start = start
+        self.end = end
+
+    def series(self, series: str):
+        for ds in self.sources:
+            maybe_ser = ds._series(series)
+            if maybe_ser:
+                series_id, series_type = maybe_ser
+                return Series.from_ch(self, ds.sess, series, series_id, series_type)
+
+    def sync(self):
+        if hasattr(self, "run_rid"):
+            return
```

Comment:
the reason I didn't do this is that this will be called recursively - so if you had 10 series in a workbook, it'd call `sync` on all of them, then each of those would call `sync` on their respective run (and eventually dataset). so the dataset/run would be uploaded the first time, then skipped for the other 9 series & produce a lot of confusing noise.

if/when we switch to a proper logger, would def add this at debug level or something